File: EdgePartition/CLUGP/src/ustc.ddl/iofiles/FileProcess.java
package ustc.ddl.iofiles;
import ustc.ddl.cluster.StreamCluster;
import ustc.ddl.properties.GlobalConfig;

import java.util.*;
import java.io.*;



public class FileProcess {
    private final  StreamCluster streamCluster;
    private final HashMap<Integer, HashMap<Integer, Integer>> innerAndCutEdge;
    private final List<Integer> clusterList;
    private final int[] cluster;



    public FileProcess(StreamCluster streamCluster) {
        this.streamCluster = streamCluster;
        this.innerAndCutEdge = streamCluster.getInnerAndCutEdge();
        this.clusterList = streamCluster.getClusterList();
        this.cluster = streamCluster.getCluster();
    }

    public void savefile() {
        try {
            BufferedWriter out = new BufferedWriter(new FileWriter(GlobalConfig.getOutputGraphPath()+"streamCluster.txt"));
            System.out.println(GlobalConfig.getOutputGraphPath()+"streamCluster.txt");

            for(int i = 0; i < cluster.length; ++i) {
                out.write(cluster[i] + " ");
            }
            out.write("\n");

            for(int i = 0; i < clusterList.size(); ++i) {
                out.write(clusterList.get(i) + " ");
            }
            out.write("\n");

            innerAndCutEdge.forEach((key1, value1) -> {
                try {
                    out.write(key1.toString());
                    out.write(" ");
                } catch (IOException ex) {
                    //exception
                }
                value1.forEach((key2, value2) ->{
                    try {
                        out.write(key2.toString());
                        out.write(" ");
                        out.write(value2.toString());
                        out.write(" ");
                    }catch (IOException ex) {
                        //exception
                    }
                });
                try {
                    out.write("\n");
                } catch (IOException ex) {
                    //exception
                }
            });
            out.close();
        } catch (IOException ex) {
            //exception
        }
    }

    public void readfile() {

        try {
            BufferedReader in = new BufferedReader(new FileReader(GlobalConfig.getOutputGraphPath()+"streamCluster.txt"));

            String line;

            line = in.readLine();
            String[] precluster = line.split(" ");

            for(int i = 0; i < precluster.length; ++i) {
                this.cluster[i] = Integer.valueOf(precluster[i]).intValue();
            }


            line = in.readLine();
            String[] preclusterList = line.split(" ");
            for(int i = 0; i < preclusterList.length; ++i) {
                this.clusterList.add(Integer.valueOf(preclusterList[i]));
            }


            while((line = in.readLine()) != null){
                String[] preinnerAndCutEdge = line.split(" ");
                //  System.out.println(preinnerAndCutEdge.length);
                HashMap<Integer, Integer> temp = new HashMap<>();
                for(int i = 1; i < preinnerAndCutEdge.length; i+=2) {
//                if (!this.innerAndCutEdge.get(Integer.valueOf(preinnerAndCutEdge[i])).containsKey(Integer.valueOf(preinnerAndCutEdge[i + 1]));
//                  innerAndCutEdge.get(cluster[src]).put(cluster[dest], 0);
                    temp.put(Integer.valueOf(preinnerAndCutEdge[i]), Integer.valueOf(preinnerAndCutEdge[i + 1]));
                }
                this.innerAndCutEdge.put(Integer.valueOf(preinnerAndCutEdge[0]), temp);
            }
            in.close();
        }catch (IOException ex) {
            //exception
        }

    }

    public StreamCluster getstreamCluster() {
        this.streamCluster.changeCluster(this.cluster);
        this.streamCluster.changeClusterList(this.clusterList);
        return this.streamCluster;
    }

}

File: EdgePartition/CLUGP/C++/FileProcess.cpp
#include "FileProcess.h"

FileProcess::FileProcess(StreamCluster* streamCluster)
	: streamCluster(streamCluster), innerAndCutEdge(streamCluster->getInnerAndCutEdge()),
		clusterList(streamCluster->getClusterList()), cluster(streamCluster->getCluster()) {}

void FileProcess::savefile(GlobalConfig& config) {
	std::string path = config.getOutputGraphPath() + "streamCluster.txt";
	std::cout << path << std::endl;

	std::ofstream out(path);
	if (!out.is_open()) {
		// Handle file open error
		std::cerr << "open output file " << path << " failed." << std::endl;
		return;
	}

	for (int i = 0; i < cluster.size(); ++i) {
		out << cluster[i] << " ";
	}
	out << "\n";

	for (int i = 0; i < clusterList.size(); ++i) {
		out << clusterList[i] << " ";
	}
	out << "\n";

	for (const auto& pair1 : innerAndCutEdge) {
		out << pair1.first << " ";
		for (const auto& pair2 : pair1.second) {
			out << pair2.first << " " << pair2.second << " ";
		}
		out << "\n";
	}
	out.close();
}

void FileProcess::readfile(GlobalConfig& config) {
	std::string path = config.getOutputGraphPath() + "streamCluster.txt";
	std::ifstream in(path);
	if (!in.is_open()) {
		std::cerr << "open input file " << path << " failed." << std::endl;
		return;
	}

	std::string line;

	// Read first line for cluster
	if (getline(in, line)) {
		std::istringstream iss(line);
		int num;
		while (iss >> num) {
			cluster.push_back(num);
		}
	}

	// Read second line for clusterList
	if (getline(in, line)) {
		std::istringstream iss(line);
		int num;
		while (iss >> num) {
			clusterList.push_back(num);
		}
	}

	// Read remaining lines for innerAndCutEdge
	while (getline(in, line)) {
		std::istringstream iss(line);
		int key1;
		iss >> key1;
		std::unordered_map<int, int> temp;
		int key2, value;
		while (iss >> key2 >> value) {
			temp[key2] = value;
		}
		innerAndCutEdge[key1] = temp;
	}
	in.close();
}

StreamCluster* FileProcess::getstreamCluster() {
	this->streamCluster->changeCluster(this->cluster);
	this->streamCluster->changeClusterList(this->clusterList);
	return this->streamCluster;
}